jlm/hls/backend/rhls2firrtl/verilator-harness-hls.cpp

```diff
@@ -1,743 +1,645 @@
   }
-  cpp << "}\n";
-
-  get_function_header(cpp, ln, "run_ref");
-  cpp << " {\n";
-  cpp << "    int fd[2]; // channel 0 for reading and 1 for writing\n"
-         "    size_t tmp = pipe(fd);\n"
-         "    int pid = fork();\n"
-         "    if(pid == 0) { // child\n"
-         "        close(fd[0]); // close fd[0] since child will only write\n"
-         "        ";
-  call_function(cpp, ln, "instrumented_ref");
-  cpp << "\n"
-         "        size_t cnt = ref_loads.size();\n"
-         "        tmp = write(fd[1], &cnt, sizeof(size_t));\n"
-         "        for (auto load:ref_loads) {\n"
-         "            tmp = write(fd[1], &load, sizeof(mem_access));\n"
-         "        }\n"
-         "        cnt = ref_stores.size();\n"
-         "        tmp = write(fd[1], &cnt, sizeof(size_t));\n"
-         "        for (auto store:ref_stores) {\n"
-         "            tmp = write(fd[1], &store, sizeof(mem_access));\n"
-         "        }\n"
-         "        close(fd[1]);\n"
-         "        exit(0);\n"
-         "    } else { // parent\n"
-         "        close(fd[1]); // close fd[1] since parent will only read\n"
-         "        size_t cnt;\n"
-         "        size_t tmp = read(fd[0], &cnt, sizeof(size_t));\n"
-         "        for (size_t i = 0; i < cnt; ++i) {\n"
-         "            mem_access load;\n"
-         "            tmp = read(fd[0], &load, sizeof(mem_access));\n"
-         "            ref_loads.push_back(load);\n"
-         "            if(load_map.find(load.addr) == load_map.end()){\n"
-         "                load_map.emplace(load.addr, std::deque<mem_access>());\n"
-         "            }\n"
-         "            load_map.find(load.addr)->second.push_back(load);\n"
-         "        }\n"
-         "        tmp = read(fd[0], &cnt, sizeof(size_t));\n"
-         "        for (size_t i = 0; i < cnt; ++i) {\n"
-         "            mem_access store;\n"
-         "            tmp = read(fd[0], &store, sizeof(mem_access));\n"
-         "            ref_stores.push_back(store);\n"
-         "            if(store_map.find(store.addr) == store_map.end()){\n"
-         "                store_map.emplace(store.addr, std::deque<mem_access>());\n"
-         "            }\n"
-         "            store_map.find(store.addr)->second.push_back(store);\n"
-         "        }\n"
-         "        close(fd[0]);\n"
-         "    }\n";
-
-  if (ln->type().NumResults() && !dynamic_cast<const rvsdg::StateType *>(&ln->type().ResultType(0)))
+  for (const auto & ctx : hls_lambda.GetContextVars())
   {
-    cpp << "    return 0;\n";
+    // Context variables should always be external symbols imported by name
+    if (const auto import = dynamic_cast<rvsdg::GraphImport *>(ctx.input->origin()))
+      cpp << "    top->i_data_" << first_ctx_var << " = (uint64_t) &" << import->Name() << ";"
+          << std::endl;
+    else
+      JLM_UNREACHABLE("Context variable is not a GraphImport");
+    first_ctx_var++;
```

Comment:
Minor nitpick and very subjetive: Invert the condtion and early check for the unreachable case, and then just put the main line code after the conditional. It would avoid the else case and the interwoven unreachable + `first_ctx_var++`

Reply:
Changed to AssertedCast instead